### DISCUSSION TITLE
2021 work to do

### DISCUSSION THREAD

**Discussion by ekexium** [tikv/client-rust#272] [General]
As we're going to release 0.1 soon, I'm trying to build a brief roadmap of the project towards its GA (1.0). There can be 3 steps:
1. Collect possible future work
2. Give each task a priority
3. Sort them and build a timeline

In terms of the future work, I have some preliminary ideas. And I'm asking for ideas from the community. Please leave the comment if you have any thought.

In my opinion, the project aims at providing a reliable, performant and easy-to-use TiKV client, so the following things can be done to help achieve the goal.

- User friendliness

  - Better error handling

    - "Drop active transaction" is sometimes annoying
    - TiKV version compatibility check. Some features require specific TiKV versions. The client should support older versions of TiKV as it evolves.

  - Better logging #267

  - More from users' feedback

  - Documentation & Example

    - I think it's good for now. We should maintain it.

  - CLI tool

    - It would be a convenient tool to explore and debug TiKV.
    - Besides usual requests, it can also provide functionalities for easier operational work (like those provided by pd-ctl or tikv-ctl)

  - We are bound to tokio now. Shall we make it compatible with any async reactor?

- Features
  - Async commit and 1PC
  - Large transaction
  - Fully functional GC (part of the TiKV GC mechanism changes)
  - Region cache
  - Parallelize multi-region request
- Reliability
  - Improve unit test
    - Currently we have low coverage (~45%, from tarpaulin)
    - We have functionalities that are not well tested, e.g. TLS
  - Improve integration test: We already have integration tests, but don't have enough coverage
    - Jepsen
    - Elle
    - Multi-region test
    - Test with nemesis or fault injection
    - Test with uncommon workload, e.g. extremely large transactions
  - Robustness
    - Don't panic if the error can be recovered by users. Only panic if it's the crate's mistake. This is done for now. We should keep it.

- Performance

  - Build an automatic benchmark script/tool
  - Provide benchmark results for each release
    - Compare with other clients if possible: java/go/python/TiDB

**Comment by ekexium** [comment]
@andylokandy @nrc Do you have anything to comment?

**Comment by nrc** [comment]
I think that our work should mostly be driven by our users post-0.1. That doesn't mean doing what people ask for or complain about, but we should find people using (or trying to use the client) and find out what they are trying to do with it. Then, we should set goals based on what people are doing. 

I think our overarching goal should be getting to 1.0. Of course, that leads to the question of what 1.0 should look like. I like your breakdown of areas we could focus on. I would break out feature completeness from reliability though. I think user friendliness and feature completeness should be our primary goals. It would be good to know what our performance constraints should be. I don't think there are issues at the moment, because what we do is so simple, but it would be good to think about some more complex cases to make sure we are doing things right. For reliability, I think we should be careful we focus on our own correctness and don't try to prove correctness of TiKV.

**Reply by ekexium** [reply]
> For reliability, I think we should be careful we focus on our own correctness and don't try to prove correctness of TiKV.

I agree. But I think end-to-end tests with simulated workload give us the best confidence in the correctness of the client. [Jepsen for TiKV](https://github.com/tikv/tikv/issues/9764) is a standalone project which uses client-rust to test TiKV.

**Comment by andylokandy** [comment]
I have some suggestions for sorting the importance:

1. The most important one is to settle down the API shape. It's very important to reach a consensus among all clients on how TiKV will be used and how features will interact with each other. By stabilizing (to some degree) the API, each client could do further internal improvement without breaking the assumption among clients. This should include the error handling, gc, large transaction, raw TTL, raw CAS etc. FYI, the project 'extract go client from TiDB' has reached the stage of refactoring user API.

2. The second important thing could be correctness. We may focus on adding more unit tests and set up the CI etc.

3. The last thing is performance improvement. This can even be done in post-1.0 because the initial users of rust client usually don't put TiKV in production, and performance improvement can be released afterward without breaking compatibility.

**Reply by ekexium** [reply]
Do you think we should try to make the API similar to (or same as) Go client's?

**Reply by andylokandy** [reply]
I'd rather like to make go client same as rust client.

**Reply by nrc** [reply]
I think that each client should have somewhat different APIs, because each should have an API which is idiomatic for the language it is written in, i.e., the Go client's API should be idiomatic Go and the Rust client's API should be idiomatic Rust. My belief is that there will not be that much overlap of users between languages, so there is not much need for uniformity - nobody will migrate their project from the Go client to the Rust client :-)

**Reply by andylokandy** [reply]
@nrc Partially agree. Your opinion is based on the idea that go client and rust client have the same feature set, and they can choose different ways to provide them. But that's not true. 

Besides, although clients should follow the language idiom in detail, they have better share the same structure. For example, `batch_get` should be at the same level as `get`, which is not true in go client.

**Reply by andylokandy** [reply]
One more thing: in go client, a key is considered not exist if the value is empty. 😄

**Reply by nrc** [reply]
> Your opinion is based on the idea that go client and rust client have the same feature set, and they can choose different ways to provide them. But that's not true.

Which part is not true?

> One more thing: in go client, a key is considered not exist if the value is empty. 😄

Oh dear, we should fix this sort of thing.

**Reply by andylokandy** [reply]
> Which part is not true?

They don't share the same feature set. Rust client supports RawTTL RawCas, and go client supports follower read, etc.

**Comment by tisonkun** [comment]
Hi guys, now we have a forum as proposed before to hold such discussion, and I've created [a `kvclient` category](https://internals.tidb.io/c/kvclient) already. Would you mind to migrate the discussion there since we do have multilingual clients and it is hard to just choose one of them and let all discussions happen in one place.

**Reply by ekexium** [reply]
I'd love to try. But I have some concerns:
1. When we discuss something specific to the rust client, is it appropriate to post it in a general client category?
2. Every contributor uses GitHub, while few of them use the internal forum. How can I notify stakeholders? Is the forum an experiment or in official operation?

**Reply by tisonkun** [reply]
This is a frequently asked questions. I think I should save the answer here :-)

> When we discuss something specific to the rust client, is it appropriate to post it in a general client category?

Nope, if you have a specific topic point to concrete code snippet change or detailed design document, file an issue in the corresponding repo or create a RFC pull request. It is mentioned in [the forum pinned](https://internals.tidb.io/t/welcome-to-tidb-internals/7):

> This forum serves for open-ended discussion of technical questions. For the concrete contribution, a contributor should file an issue or follow the RFC process.

So, it is for open-ended discussion such as a code snippet should be refactor(but unsure about the impact), a series of logics may be problematic, a rough idea of new feature, or a plan to decouple code and formalize interface.

For example, this post is about the roadmap of the whole client project, and thus is a typical open-ended topic to discuss on a forum. It is an invitation for you experts to discuss on the forum and share audiences of TiDB projects.

> How can I notify stakeholders? Is the forum an experiment or in official operation?

As mentioned above, for the development procedure, we have dedicated issue & pull request process for the actual development task. For an official channel for announcement or decision making, I tend to gather all technical decision making under https://github.com/tikv/rfcs. But as offline discussion with @nrc about how Rust teams ever did, it is possibly you guys define your own way for decision making on the top of issue & pull request.

Now it is this Discussion and ever it is sig-transaction repo. I'm the person in charge of TiDB Internals and invite you to discuss on the forum and share audiences of TiDB projects. I believe it would be a good channel especially for contributor marketing under the same interface.

**Reply by ekexium** [reply]
Got it, thank you!

> For example, this post is about the roadmap of the whole client project

This is for the rust client only, so I think the topic should stay here. Though I'm hesitating whether it should be an issue or a discussion.

**Reply by tisonkun** [reply]
> This is for the rust client only, so I think the topic should stay here. Though I'm hesitating whether it should be an issue or a discussion.

@ekexium ok, I get it. Given that we have TiDB Internals forum for open-ended discussion, I'd prefer move discussion there and supersede GitHub Discussion. We don't have to maintain two channels though. For a specific roadmap of a project, it can be an issue if you think it is detailed enough or just like it, like [ClickHouse](https://github.com/ClickHouse/ClickHouse/issues/17623) since a discussion cannot be converted to an issue. But it is up to you after all.

**Reply by tisonkun** [reply]
Besides, I suggest to use a title RoadMap for Rust Client 2021 instead of Q2/Q3 which looks like an enterprise private concept.

**Reply by ekexium** [reply]
Yeah, it's redundant to maintain too many places for discussion. If we're going to use the internal forum to discuss client-rust specific content, would it be better to have a separate category, or just use the kvclient one?

This thread only discusses possible work in the next months. I'm drafting a roadmap which also includes priorities and a timeline.

**Reply by tisonkun** [reply]
I think we don't create a category with an over subdivided name. I don't use `client` because it is too general to be regraded as MySQL Client / MPPClient / KVClient and so on. `kvclient` is specific and clear for TiKV Client, distinguish the server side. I think the Go Client and the Rust Client are the most active developing ones and able to share the same concept at the moment.

**Reply by ekexium** [reply]
All right. Let's keep it simple for now and see how it works.

**Comment by nrc** [comment]
Another thing to consider is if/how we should provide a non-async API

**Reply by ekexium** [reply]
Why should we provide non-async API?

**Reply by nrc** [reply]
Wrapping async to provide sync is very un-ergonomic and not performant, and for many users, using traditional threads is a better design choice than using async

**Reply by nrc** [reply]
See for example MongoDb's client which supports async with Tokio or async-std, or sync: https://github.com/mongodb/mongo-rust-driver

**Reply by ekexium** [reply]
It seems like MongoDB's sync client just `block_on` each async operation.

**Reply by nrc** [reply]
We could do that to start with, but should have something better for the long term

**Reply by andylokandy** [reply]
@nrc I think before anyone complains about the sync not available we don't need to care about it.

**Reply by nrc** [reply]
The thing about sync is that it is easier to get started, so I think that if we don't have it, people are more likely to just not try out the client, rather than complain. It should be very easy to provide a not very efficient sync API just by wrapping everything in `block_on`
